certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,401 @@
+invariant countCorrect()
+    forall address wallet. forall address X. (ghostSuccCount(wallet, X) == countExpected(wallet, X)) && countSuccessor(wallet, X)
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant inListReachable();
+            requireInvariant reachNext();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+            requireInvariant countZeroIffListEmpty();
+        }
+    }
+
+// Invariant that checks for any wallet either there are no guardians (and SENTINEL points to NULL) or successor count of SENTINEL equals ghostGuardianCount + 1.
+invariant guardianCountCorrect()
+   forall address wallet. (ghostGuardianCount[wallet] == 0 && ghostGuardians[wallet][SENTINEL] == NULL) || (ghostSuccCount(wallet, SENTINEL) == ghostGuardianCount[wallet] + 1)
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant inListReachable();
+            requireInvariant reachNext();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+        }
+    }
+
+// The ghostGuardians[wallet][SENTINEL] should be NULL or SENTINEL if the guardian count is 0.
+invariant countZeroIffListEmpty()
+    forall address wallet. ghostGuardianCount[wallet] == 0 <=>
+        (ghostGuardians[wallet][SENTINEL] == NULL || ghostGuardians[wallet][SENTINEL] == SENTINEL)
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant inListReachable();
+            requireInvariant reachNext();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+            requireInvariant countCorrect();
+            requireInvariant guardianCountCorrect();
+        }
+    }
+
+// All the elements in ghostGuardians[wallet] should point to NULL if the list is empty.
+invariant emptyListNotReachable()
+    forall address wallet. (ghostGuardians[wallet][SENTINEL] == NULL || ghostGuardians[wallet][SENTINEL] == SENTINEL)
+        => (forall address X. X != SENTINEL => ghostGuardians[wallet][X] == NULL)
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant reachNull();
+            requireInvariant inListReachable();
+            requireInvariant reachNext();
+            requireInvariant nextNull();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+            requireInvariant countCorrect();
+            requireInvariant guardianCountCorrect();
+        }
+    }
+
+rule storeHookPreservesInvariants(address wallet, address key, address value) {
```

Comment:
Should we have some comment for this one?

Reply:
@remedcu thanks for suggesstions. I will add them in another PR

Reply:
Sure, didn't saw it was merged until it was too late 😅